packages/ifpack2/src/Ifpack2_Details_CrsArrays.hpp

```diff
@@ -0,0 +1,187 @@
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ***********************************************************************
+//@HEADER
+*/
+
+#ifndef __IFPACK2_CRSARRAYS_DECL_HPP__ 
+#define __IFPACK2_CRSARRAYS_DECL_HPP__ 
+
+#include <Tpetra_RowMatrix.hpp>
+#include <Tpetra_CrsMatrix.hpp>
+#include <Kokkos_DefaultNode.hpp>
+#include <Kokkos_CrsMatrix.hpp>
+#include <Ifpack2_LocalFilter.hpp>
+#include <Ifpack2_ReorderFilter.hpp>
+
+namespace Ifpack2
+{
+namespace Details
+{
+
+//Utility for getting the local values, rowptrs and colinds (in Kokkos::Views) for any RowMatrix
+//Used by Fic, Filu and Fildl but may also be useful in other classes
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+struct CrsArrayReader
+{
+  typedef typename Node::device_type Device;
+  typedef typename Device::execution_space ExecSpace;
+  typedef Tpetra::RowMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> TRowMatrix;
+  typedef Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> TCrsMatrix;
+  typedef Ifpack2::LocalFilter<TRowMatrix> Filter;
+  typedef Ifpack2::ReorderFilter<TRowMatrix> ReordFilter;
+  typedef Kokkos::CrsMatrix<Scalar, LocalOrdinal, ExecSpace> KCrsMatrix;
+  typedef Kokkos::View<LocalOrdinal*, ExecSpace> OrdinalArray;
+  typedef Kokkos::View<Scalar*, ExecSpace> ScalarArray;
+
+  //These functions are only meant to be used by Fic/Filu/Fildl for AdditiveSchwarz
+  //A's underlying type can either be a CrsFilter or a RowFilter
+  static void getValues(const TRowMatrix* A, ScalarArray& vals)
+  {
+    auto Acrs = dynamic_cast<const TCrsMatrix*>(A);
+    if(Acrs)
+    {
+      getValuesCrs(Acrs, vals);
+      return;
+    }
+    //Need to allocate new array, then copy in one row at a time
+    vals = ScalarArray("", A->getNodeNumEntries());
+    //is there a better lower bound available for this, without allocating a new array for each row?
+    //RowMatrix doesn't have "getMaxNnzPerRow", but that's really what this should be
+    std::vector<Scalar> valsVec(A->getNodeMaxNumRowEntries());
+    Teuchos::ArrayView<Scalar> valsView(valsVec);
+    std::vector<LocalOrdinal> indicesVec(A->getNodeMaxNumRowEntries());
+    Teuchos::ArrayView<LocalOrdinal> indicesView(indicesVec);
+    size_t entry = 0;
+    for(LocalOrdinal row = 0; row < A->getNodeNumRows(); row++)
+    {
+      //Make an ArrayView that views this row's values within vals
+      auto rowEntries = A->getNumEntriesInLocalRow(row);
+      A->getLocalRowCopy(row, indicesView, valsView, rowEntries);
+      //sort the indices, permuting values at the same time
+      Tpetra::sort2(indicesView.getRawPtr(), indicesView.getRawPtr() + rowEntries, valsView.getRawPtr());
+      for(LocalOrdinal i = 0; i < rowEntries; i++)
+      {
+        vals(entry++) = valsView[i];
```

Comment:
On the other hand, you could get the same effect just by bulking up the copies into larger host chunks before sending them to device.  It might not be worthwhile, given that you're copying stuff anyway.